File: tools/source_editing/review_lint/api/scan_diff.py
#!/usr/bin/env python3
"""Diff-aware QA ship gate: scan a unified diff for maintainer-rejected patterns.

Runs the review_lint QA rules (extern-literal anchors, string-literal-to-
symbol swaps, packed string blobs, unrolled asserts, banned patterns,
resubmission tombstones) against the ADDED lines of a diff only, so
pre-existing upstream code is never flagged.

Output contract (mirrors packages/core/src/qa/scan-diff.ts):
  stdout: JSON {tool, operation, status, repo, base, findings, counts}
  stderr: human-readable summary
  exit (with --gate): 0 clean, 1 any error finding, 2 warnings only.
  Without --gate the exit code is always 0 unless the tool itself fails.
"""

from __future__ import annotations

import argparse
import json
import re
import subprocess
import sys
from pathlib import Path
from typing import Any

sys.path.append(str(Path(__file__).resolve().parents[3] / "_shared"))
sys.path.insert(0, str(Path(__file__).resolve().parent))

import _qa_rules
import check_extern_ownership

HUNK_HEADER_RE = re.compile(r"^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@")
GLOBAL_APPLIES_TO = ["src/**/*.c"]


def run_git(repo: Path, args: list[str]) -> str:
    """Run a read-only git command in the target repo and return stdout."""

    result = subprocess.run(
        ["git", "-C", str(repo), *args],
        capture_output=True,
        text=True,
        check=False,
    )
    if result.returncode != 0:
        raise RuntimeError(
            f"git {' '.join(args)} failed (exit {result.returncode}): "
            f"{result.stderr.strip()}"
        )
    return result.stdout


def parse_unified_diff(diff_text: str) -> list[dict[str, Any]]:
    """Parse a unified diff into per-file hunk records.

    Returns [{"file": path, "hunks": [{"file", "added", "removed", "context"},
    ...]}]. Added lines carry their new-file line numbers; removed and
    context lines are bare text (both existed in the base version).
    """

    files: list[dict[str, Any]] = []
    current_file: dict[str, Any] | None = None
    hunk: dict[str, Any] | None = None
    new_lineno = 0
    for raw in diff_text.splitlines():
        if raw.startswith("+++ "):
            path = raw[4:].strip()
            if path.startswith("b/"):
                path = path[2:]
            if path == "/dev/null":
                current_file = None
            else:
                current_file = {"file": path, "hunks": []}
                files.append(current_file)
            hunk = None
            continue
        if raw.startswith("--- ") or raw.startswith("diff --git") or raw.startswith("index "):
            hunk = None
            continue
        header = HUNK_HEADER_RE.match(raw)
        if header:
            if current_file is None:
                hunk = None
                continue
            new_lineno = int(header.group(3))
            hunk = {
                "file": current_file["file"],
                "added": [],
                "removed": [],
                "context": [],
            }
            current_file["hunks"].append(hunk)
            continue
        if hunk is None:
            continue
        if raw.startswith("\\"):
            continue  # "\ No newline at end of file"
        if raw.startswith("+"):
            hunk["added"].append((new_lineno, raw[1:]))
            new_lineno += 1
        elif raw.startswith("-"):
            hunk["removed"].append(raw[1:])
        else:
            # Context line (leading space; a fully blank line also counts).
            hunk["context"].append(raw[1:] if raw.startswith(" ") else raw)
            new_lineno += 1
    return [record for record in files if record["hunks"]]


def post_diff_file_text(
    repo: Path,
    rel_path: str,
    mode: str,
    file_diffs: list[dict[str, Any]],
) -> str | None:
    """Return post-diff text used for the in-file-definition check.

    - "head": ref-mode diff against HEAD -> `git show HEAD:<file>`.
    - "worktree": --include-worktree -> read the worktree file.
    - "diff": --diff-file mode -> the diff's own added lines for the file.
      The worktree is NOT consulted here: it may sit on an unrelated branch,
      and a moved definition (the legitimate ftcoll case) is visible in the
      added lines themselves.
    """

    if mode == "worktree":
        path = repo / rel_path
        if path.is_file():
            return path.read_text(encoding="utf-8", errors="replace")
        return None
    if mode == "diff":
        lines: list[str] = []
        for record in file_diffs:
            if record["file"] != rel_path:
                continue
            for hunk in record["hunks"]:
                lines.extend(text for _, text in hunk["added"])
        return "\n".join(lines)
    try:
        return run_git(repo, ["show", f"HEAD:{rel_path}"])
    except RuntimeError:
        return None


def has_in_file_definition(file_text: str, symbol: str) -> bool:
    """Check whether the TU defines (not just declares extern) the symbol."""

    clean = _qa_rules.strip_comments_and_strings(file_text)
    name = re.escape(symbol)
    init_re = re.compile(
        r"^\s*(?:static\s+)?(?:(?:const|volatile)\s+)*"
        r"(?:f32|f64|float|double|char|u8|s8|u16|s16|u32|s32|int|unsigned|signed|long|short)\b"
        rf"[\w \t*]*?\b{name}\s*(?:\[[^\]]*\])?\s*(?:=(?!=)|;)"
    )
    for line in clean.splitlines():
        if "extern" in line:
            continue
        if init_re.match(line):
            return True
    return False


def symbol_in_diff_base(
    file_diffs: list[dict[str, Any]], rel_path: str, symbol: str
) -> bool:
    """Infer from the diff itself whether a symbol existed in the BASE file.

    A symbol appearing in any removed (-) or context (space) line of the
    file's hunks must have existed in the base version; a symbol appearing
    only in added (+) lines is new in this diff.

    Caveat: unified-diff context is partial (only a few lines around each
    change), so a base symbol that the diff never touches or passes near is
    invisible here and would be misread as new. For the cases this gate
    targets the inference holds: moving a pre-existing definition later in
    the file (the accepted ftcoll pattern) necessarily shows the old
    definition as removed lines, while an invented anchor (the rejected
    gm_1832 pattern) appears only in added lines.
    """

    pattern = re.compile(rf"\b{re.escape(symbol)}\b")
    for record in file_diffs:
        if record["file"] != rel_path:
            continue
        for hunk in record["hunks"]:
            for text in hunk["removed"]:
                if pattern.search(text):
                    return True
            for text in hunk.get("context", []):
                if pattern.search(text):
                    return True
    return False


def symbol_existed_in_base(
    repo: Path,
    rel_path: str,
    symbol: str,
    mode: str,
    file_diffs: list[dict[str, Any]],
    merge_base: str | None,
    base_text_cache: dict[str, str | None],
) -> bool:
    """Determine whether a symbol existed in the base version of a file.

    Ref mode prefers `git show <merge-base>:<file>` (authoritative full base
    text); --diff-file mode (and a missing base blob) falls back to the
    removed/context-line inference from the diff itself.
    """

    if mode != "diff" and merge_base:
        if rel_path not in base_text_cache:
            try:
                base_text_cache[rel_path] = run_git(
                    repo, ["show", f"{merge_base}:{rel_path}"]
                )
            except RuntimeError:
                base_text_cache[rel_path] = None  # new file in this diff
        base_text = base_text_cache[rel_path]
        if base_text is not None:
            return re.search(rf"\b{re.escape(symbol)}\b", base_text) is not None
    return symbol_in_diff_base(file_diffs, rel_path, symbol)


def evaluate_extern_findings(
    findings: list[dict[str, Any]],
    repo: Path,
    mode: str,
    file_diffs: list[dict[str, Any]],
    merge_base: str | None = None,
) -> list[dict[str, Any]]:
    """Apply ownership analysis to extern_literal_anchor float warnings.

    - definition elsewhere in the same TU AND the symbol existed in the base
      file -> forward declaration of moved data (accepted ftcoll style), drop.
    - definition elsewhere in the same TU but the symbol is entirely new in
      this diff -> new_data_anchor error (the extern + brand-new definition
      invents a data anchor to force data ordering; rejected gm_1832 style).
    - no in-TU definition, encoded address inside the TU's own data ranges
      -> self_tu_extern error.
    - otherwise -> legitimate cross-TU reference, keep as warning.
    """

    file_text_cache: dict[str, str | None] = {}
    base_text_cache: dict[str, str | None] = {}
    result: list[dict[str, Any]] = []
    for finding in findings:
        if finding["rule_id"] != "extern_literal_anchor":
            result.append(finding)
            continue
        detail = finding.get("detail") or {}
        ctype = detail.get("ctype")
        if ctype not in _qa_rules.FLOAT_TYPES:
            result.append(finding)
            continue
        rel_path = finding["file"]
        symbol = detail.get("symbol", "")
        if rel_path not in file_text_cache:
            file_text_cache[rel_path] = post_diff_file_text(
                repo, rel_path, mode, file_diffs
            )
        file_text = file_text_cache[rel_path]
        if file_text and symbol and has_in_file_definition(file_text, symbol):
            if symbol_existed_in_base(
                repo, rel_path, symbol, mode, file_diffs, merge_base, base_text_cache
            ):
                # Forward declaration of pre-existing data the TU still
                # defines later (definition moved within the file): clean.
                continue
            # The extern, its uses, AND the definition are all new in this
            # diff: an invented data anchor to force data ordering.
            standard = "global_standard:literals-and-data-ownership"
            finding = dict(finding)
            finding["rule_id"] = "new_data_anchor"
            finding["severity"] = "error"
            finding["standard_id"] = standard
            finding["message"] = (
                f"`extern {ctype} {symbol}` introduces a brand-new data anchor: "
                "the symbol did not exist in the base file and this diff adds "
                "the extern, its use, and the definition together. Do not "
                '"Create static literals or globals solely to force data order" '
                '— "using an extern to make a function match is just due to '
                'data ordering". '
                f"{_qa_rules.STANDARD_TITLES[standard]}."
            )
            finding["detail"] = {**detail, "verdict": "new_data_anchor"}
            result.append(finding)
            continue
        address = _qa_rules.address_from_name(symbol)
        if address is not None and check_extern_ownership.tu_owns_address(
            repo, rel_path, address
        ):
            finding = dict(finding)
            finding["rule_id"] = "self_tu_extern"
            finding["severity"] = "error"
            finding["message"] = (
                f"`extern {ctype} {symbol}` references data this TU owns "
                f"(0x{address:08X} is inside its own section ranges in splits.txt). "
                "Externing your own data dodges data ordering; define the constant "
                "in binary order instead. "
                f"{_qa_rules.STANDARD_TITLES['global_standard:literals-and-data-ownership']}."
            )
            finding["detail"] = {**detail, "verdict": "confirmed_self_tu"}
            result.append(finding)
        else:
            finding = dict(finding)
            finding["detail"] = {**detail, "verdict": "cross_tu_ok"}
            result.append(finding)
    return result


def collect_findings(
    file_diffs: list[dict[str, Any]],
    repo: Path,
    mode: str,
    merge_base: str | None = None,
) -> list[dict[str, Any]]:
    """Run all rules (built-ins, banned patterns, tombstones) over the diff."""

    rules = _qa_rules.all_rules(include_banned=True)
    tombstones = _qa_rules.load_tombstones()
    findings: list[dict[str, Any]] = []
    for record in file_diffs:
        if not _qa_rules.path_matches(record["file"], GLOBAL_APPLIES_TO):
            continue
        for hunk in record["hunks"]:
            findings.extend(_qa_rules.run_rules_on_hunk(rules, hunk))
            for partial in _qa_rules.check_tombstones(hunk, tombstones):
                finding = {"file": record["file"], **partial}
                finding["excerpt"] = finding["excerpt"][:240]
                findings.append(finding)
    findings = evaluate_extern_findings(findings, repo, mode, file_diffs, merge_base)
    findings.sort(key=lambda f: (f["file"], f["line"], f["rule_id"]))
    return findings


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--repo", required=True, help="Melee repo root.")
    parser.add_argument(
        "--base",
        default=None,
        help="Base ref to diff against (merge-base with HEAD; default origin/master).",
    )
    parser.add_argument(
        "--diff-file",
        default=None,
        help="Pre-computed unified diff to scan instead of a ref diff.",
    )
    parser.add_argument(
        "--path",
        action="append",
        default=[],
        help="Restrict the ref diff to this repo-relative pathspec (repeatable).",
    )
    parser.add_argument(
        "--include-worktree",
        action="store_true",
        help="Diff the worktree against the merge-base instead of HEAD.",
    )
    parser.add_argument(
        "--gate",
        action="store_true",
        help="Exit 1 on error findings, 2 on warnings only, 0 when clean.",
    )
    parser.add_argument(
        "--json", action="store_true", help="Accepted for symmetry; JSON is always emitted."
    )
    args = parser.parse_args()

    if args.diff_file and args.base:
        parser.error("--diff-file and --base are mutually exclusive")

    repo = Path(args.repo).expanduser().resolve()
    if not repo.is_dir():
        print(f"scan_diff: repo not found: {repo}", file=sys.stderr)
        return 3

    merge_base: str | None = None
    mode = "worktree" if args.include_worktree else "head"
    try:
        if args.diff_file:
            diff_path = Path(args.diff_file)
            if not diff_path.is_file():
                print(f"scan_diff: diff file not found: {diff_path}", file=sys.stderr)
                return 3
            diff_text = diff_path.read_text(encoding="utf-8", errors="replace")
            # In --diff-file mode the post-diff state is reconstructed from
            # the diff's own added lines (the repo worktree may be on an
            # unrelated branch).
            mode = "diff"
        else:
            base_ref = args.base or "origin/master"
            merge_base = run_git(repo, ["merge-base", base_ref, "HEAD"]).strip()
            pathspecs = args.path or ["src"]
            diff_args = ["diff", "--no-color", "--unified=5", merge_base]
            if not args.include_worktree:
                diff_args.append("HEAD")
            diff_args.append("--")
            diff_args.extend(pathspecs)
            diff_text = run_git(repo, diff_args)
    except RuntimeError as error:
        print(f"scan_diff: {error}", file=sys.stderr)
        return 3

    file_diffs = parse_unified_diff(diff_text)
    if args.diff_file and args.path:
        wanted = set(args.path)
        file_diffs = [
            record
            for record in file_diffs
            if record["file"] in wanted
            or any(record["file"].startswith(p.rstrip("/") + "/") for p in wanted)
        ]
    findings = collect_findings(file_diffs, repo, mode, merge_base)

    errors = sum(1 for f in findings if f["severity"] == "error")
    warnings = sum(1 for f in findings if f["severity"] == "warning")
    status = "failed" if errors else ("warned" if warnings else "passed")
    payload = {
        "tool": "review_lint",
        "operation": "review_lint:scan_diff",
        "status": status,
        "repo": str(repo),
        "base": merge_base,
        "findings": findings,
        "counts": {"errors": errors, "warnings": warnings},
    }
    print(json.dumps(payload, indent=2, sort_keys=True))

    print(
        f"review_lint scan_diff: {status} "
        f"({errors} error(s), {warnings} warning(s), "
        f"{len(file_diffs)} scanned file(s))",
        file=sys.stderr,
    )
    for finding in findings:
        print(
            f"  [{finding['severity']}] {finding['rule_id']} "
            f"{finding['file']}:{finding['line']} — {finding['message']}",
            file=sys.stderr,
        )

    if args.gate:
        if errors:
            return 1
        if warnings:
            return 2
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: apps/server/src/core/validation/qa/scan-diff.ts
/**
 * Shared invoker for the review_lint diff-aware QA scanner.
 *
 * The scanner is the deterministic layer of the QA ship gate: it runs the
 * maintainer-rejection rules (literal/data-symbol substitutions, packed string
 * blobs, copied header inlines, stage GroundVars ownership, unrolled asserts,
 * banned patterns, resubmission tombstones) against the added lines of a diff.
 * Both the worker-side L1 check and the
 * regression-check L2 ship gate go through this helper so they share one
 * contract.
 *
 * Exit-code contract for `scan_diff.py --gate`:
 *   0 = clean, 1 = hard-fail findings present, 2 = warnings only.
 * Stdout is always the JSON document; the human summary goes to stderr.
 */
import { existsSync } from "node:fs";
import { spawn } from "node:child_process";
import { resolve } from "node:path";
import type { RunProjectMetadata } from "@server/core/shared/types";
import { resolveRegisteredTool } from "@server/core/tools/resolver";

export type QaScanSeverity = "error" | "warning";

export interface QaScanFinding {
  rule_id: string;
  severity: QaScanSeverity;
  file: string;
  line: number;
  excerpt: string;
  message: string;
  standard_id: string | null;
  /** Extra rule-specific context (ownership verdicts, tombstone refs, ...). */
  detail?: Record<string, unknown>;
}

export interface QaScanResult {
  tool: "review_lint";
  operation: "review_lint:scan_diff";
  status: "passed" | "warned" | "failed";
  repo: string;
  base: string | null;
  findings: QaScanFinding[];
  counts: { errors: number; warnings: number };
}

export interface QaScanInvocation {
  /** Exit code from scan_diff.py (0 clean / 1 hard fail / 2 warnings; other = tool failure). */
  exitCode: number;
  /** Parsed stdout JSON, or null when stdout was not parseable. */
  result: QaScanResult | null;
  /** Raw stdout/stderr for artifact capture. */
  stdout: string;
  stderr: string;
  /** Set when the tool itself failed (script missing, crash, bad JSON). */
  toolError: string | null;
  command: string[];
}

export interface RunQaScanDiffOptions {
  /** Colosseum (target project) repo root the diff lives in. */
  repoRoot: string;
  /** Orchestrator root. Kept for compatibility with older callers. */
  orchestratorRoot: string;
  /** Project metadata used to resolve project tool bindings when available. */
  project?: RunProjectMetadata;
  /** Project state dir used to resolve tool cache/worktree roots when available. */
  stateDir?: string;
  /** Explicit worktree id for parallel validation worktrees. */
  worktreeId?: string;
  /** Base ref to diff against (merge-base is computed by the tool). Mutually exclusive with diffFile/files. */
  baseRef?: string;
  /** Pre-computed unified diff file to scan instead of a ref diff. */
  diffFile?: string;
  /** Restrict the ref diff to these pathspecs (worker L1 scoping). */
  files?: string[];
  /** Include uncommitted worktree edits in ref-mode scans. */
  includeWorktree?: boolean;
  /** Run scanner in gate mode. Defaults to true. Queue-building scans can disable this to collect findings without a failing process exit. */
  gate?: boolean;
}

async function runProcess(repoRoot: string, command: string[], env?: Record<string, string>): Promise<{ exitCode: number; stdout: string; stderr: string }> {
  return new Promise((resolveProcess) => {
    const child = spawn(command[0] ?? "", command.slice(1), { cwd: repoRoot, env: env ? { ...process.env, ...env } : process.env });
    const stdout: Buffer[] = [];
    const stderr: Buffer[] = [];
    let closed = false;
    let stdoutEnded = false;
    let stderrEnded = false;
    let exitCode = -1;
    let spawnError = "";
    const finish = () => {
      if (!closed || !stdoutEnded || !stderrEnded) return;
      resolveProcess({
        exitCode,
        stdout: Buffer.concat(stdout).toString("utf8"),
        stderr: `${Buffer.concat(stderr).toString("utf8")}${spawnError}`,
      });
    };
    child.stdout.on("data", (chunk: Buffer) => stdout.push(chunk));
    child.stdout.on("end", () => {
      stdoutEnded = true;
      finish();
    });
    child.stderr.on("data", (chunk: Buffer) => stderr.push(chunk));
    child.stderr.on("end", () => {
      stderrEnded = true;
      finish();
    });
    child.on("error", (error) => {
      spawnError = error.message;
      exitCode = -1;
      stdoutEnded = true;
      stderrEnded = true;
      closed = true;
      finish();
    });
    child.on("close", (code) => {
      exitCode = code ?? -1;
      closed = true;
      finish();
    });
  });
}

export function qaScanDiffScriptPath(orchestratorRoot: string): string {
  return resolve(orchestratorRoot, "toolpacks/gamecube-decomp/source_editing/review_lint/api/scan_diff.py");
}

export function qaGatePassed(invocation: QaScanInvocation): boolean {
  return invocation.toolError === null && invocation.result !== null && invocation.result.counts.errors === 0 && invocation.result.counts.warnings === 0 && invocation.exitCode === 0;
}

export function parseQaScanResult(stdout: string): QaScanResult | null {
  try {
    const parsed = JSON.parse(stdout) as Record<string, unknown>;
    if (parsed && parsed.tool === "review_lint" && Array.isArray(parsed.findings)) {
      return parsed as unknown as QaScanResult;
    }
    return null;
  } catch {
    return null;
  }
}

function cleanResultFromExitZero(options: RunQaScanDiffOptions, stderr: string): QaScanResult | null {
  if (stderr.trim() && !stderr.match(/review_lint scan_diff: passed \(0 error\(s\), 0 warning\(s\), \d+ scanned file\(s\)\)/)) return null;
  return {
    tool: "review_lint",
    operation: "review_lint:scan_diff",
    status: "passed",
    repo: options.repoRoot,
    base: options.baseRef ?? null,
    findings: [],
    counts: { errors: 0, warnings: 0 },
  };
}

export async function runQaScanDiff(options: RunQaScanDiffOptions): Promise<QaScanInvocation> {
  const resolved = resolveRegisteredTool(
    {
      project: options.project,
      repoRoot: options.repoRoot,
      stateDir: options.stateDir,
      worktreeId: options.worktreeId,
    },
    "review_lint",
  );
  const scriptPath = resolve(resolved.apiRoot, "scan_diff.py");
  const command = ["python3", scriptPath, "--repo", options.repoRoot, "--json"];
  if (options.gate !== false) command.push("--gate");
  if (options.baseRef) command.push("--base", options.baseRef);
  if (options.diffFile) command.push("--diff-file", options.diffFile);
  if (options.includeWorktree) command.push("--include-worktree");
  for (const file of options.files ?? []) command.push("--path", file);
  if (!existsSync(scriptPath)) {
    return {
      exitCode: -1,
      result: null,
      stdout: "",
      stderr: "",
      toolError: `scan_diff.py not found at ${scriptPath}`,
      command,
    };
  }
  const result = await runProcess(options.repoRoot, command, resolved.env);
  const parsed = parseQaScanResult(result.stdout) ?? (result.exitCode === 0 && result.stdout.trim() === "" ? cleanResultFromExitZero(options, result.stderr) : null);
  const toolError =
    parsed === null
      ? `scan_diff.py did not return parseable JSON (exit ${result.exitCode})`
      : ![0, 1, 2].includes(result.exitCode)
        ? `scan_diff.py failed with exit ${result.exitCode}`
        : null;
  return {
    exitCode: result.exitCode,
    result: parsed,
    stdout: result.stdout,
    stderr: result.stderr,
    toolError,
    command,
  };
}
